_posts/2019-06-04-loading-large-image-data-with-dask-array.md

```diff
@@ -0,0 +1,150 @@
+process.
+
+Historically this analytical process has involved a lot of custom code. Often
+the analytical process is stitched together by a series of scripts possibly in
+several different languages that write various intermediate results to disk.
+Thanks to advances in modern tooling these process can be significantly
+improved. In this series of blogposts, we will outline ways for image
+scientists to leverage different tools to move towards a high level, friendly,
+cohesive, interactive analytical pipeline.
+
+Executive Summary
+-----------------
+
+This post explores a few simple workflows that a user might leverage to load
+their image with Dask. At the end of this post users should be able to
+construct a Dask Array, which they can use to further explore and manipulate
+their image data.
+
+Motivation
+----------
+
+Loading large image data can be a complex and often unique problem. Different
+groups may choose to store this across many files on disk, a commodity or
+custom database solution, or they may opt to store it in the cloud. Not all
+datasets within the same group may be treated the same for a variety of
+reasons. In short, this means loading data is a hard and expensive problem.
+
+Despite data being stored in different ways, often group want to reapply the
+same analytical pipeline to these datasets. However if the data pipeline is
+tightly coupled to a particular way of loading the data for later analytical
+steps, it may be very difficult if not impossible to reuse an existing
+pipeline. In other words, there is friction between the loading and analysis
+steps, which frustrates efforts to make things reusable.
+
+Having a modular and general way to load data makes it easy to present data
+stored differently in a standard way. Further having a standard way to present
+data to analytical pipelines allows that part of the pipeline to focus on what
+it does best, analysis! In general, this should decouple these to components in
+a way that improves the experience of users involved in all parts of the
+pipeline.
+
+
+Loading image data with Dask
+----------------------------
+
+A pretty typical case to see is a large directory of image files. In particular
+TIFF files are quite common, though other formats could show up as well. It is
+common to see a naming structure that mirrors that shown below. The `<i>`
+represent different numbers.
+
+```
+mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif
+```
+
+To load a single one of these a user might right some code like this:
+
+```python
+from skimage.io import imread
+a = imread("mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif")
+```
+
+While this works fine for reading a single image into memory, this will quickly
+become prohibitive once we try to load more data than we have memory. So we
+might replace this with something like this.
+
+```python
+from dask.array.image import imread
+a = imread("mydata_ch<i>_<j>t_<k>x_<l>y_<m>z.tif")
+```
```

Comment:
Thoughts on starting with dask delayed instead?

Reply:
Yep, we could do that. Should we keep this code or merely reference it in text?

Reply:
I've added some code before this showing delayed. Have also reworked this section to match [this outline]( https://github.com/dask/dask-blog/pull/30#discussion_r292674949 ).

Reply:
If we have uniform shapes and dtypes then I think that there is a more effective pattern than what is shown above. 

```python
sample = skimage.io.imread(filenames[0])
```

Show output of single sample image


One sentence about how we can now lazily read in all of the files as lazy single-block dask arrays, using the sample as a pattern.

```python
arrays = [da.from_delayed(dask.delayed(skimage.io.imread)(fn), shape=sample.shape, dtype=sample.dtype) for fn in filenames]
```

One or two sentences about how we can then concatenate or stack these arrays into a larger dask array, followed by tiny code examples with da.concatenate and da.stack and their visualized shapes